File: app/models/combined_article.rb
class CombinedArticle
  include ActiveModel::Model

  def initialize(ids)
    articles = Article.where(id: ids)
    @a = articles[0]
    @b = articles[1]
    @c = articles[2]
  end

  def title
    @a.title
  end

  def sections
    secs = @a.headers
    steps_per = steps.length / secs.length
    steps_ary = steps
    images_ary = images

    secs.map.with_index do |section, i|
      {
        number: i + 1,
        header: section,
        steps: Array.new(steps_per.to_i) do |i|
          {
            number: i + 1,
            content: steps_ary.pop,
            image: images_ary.pop
          }
        end
      }
    end
  end

  def steps
    @steps ||= @b.steps
  end

  def images
    @images ||= @c.images
  end

end

File: frontend/src/models/CombinedArticle.js
class CombinedArticle {

  constructor(articles) {
    this.a = articles[0];
    this.b = articles[1];
    this.c = articles[2];

    if (!this.a || !this.b || !this.c) {
      this.failed = true;
    }
  }

  get title() {
    return this.a.title;
  }

  get sections() {
    return this.a.sections;
  }

  get steps() {
    return this.b.steps;
  }

  get images() {
    return this.c.images;
  }

  get content() {
    if (this.failed) { return []; }

    let secs = this.sections;
    let steps = this.steps;
    let images = this.images;

    const secsCount = Math.max(secs.length, 1);
    const stepsPer = Math.max(Math.round(steps.length / secsCount), 1);
    const ary = Array.from({length: stepsPer}, (v, i) => i);

    return secs.map((section, si) => {
      return {
        number: si + 1,
        header: section,
        steps: ary.map((i) => {
          return {
            number: i + 1,
            content: steps.pop(),
            image: images.pop()
          }
        })
      }
    });
  }
}

export default CombinedArticle;
